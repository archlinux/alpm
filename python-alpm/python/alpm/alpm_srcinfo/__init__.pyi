"""A module for parsing and linting of ALPM SRCINFO files."""

from . import error, source_info
from .error import SourceInfoError
from .source_info.v1 import SourceInfoV1
from .source_info.v1.merged import MergedPackage

__all__ = [
    "SourceInfoError",
    "error",
    "source_info",
    "SourceInfoV1",
    "MergedPackage",
]
