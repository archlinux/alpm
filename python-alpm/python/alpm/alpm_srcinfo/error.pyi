"""Exceptions used by alpm_srcinfo module."""

class SourceInfoError(Exception):
    """The high-level exception that can occur when using alpm_srcinfo module."""

__all__ = [
    "SourceInfoError",
]
