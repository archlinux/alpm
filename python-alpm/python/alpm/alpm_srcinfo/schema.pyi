"""Schemas for SRCINFO data."""

from pathlib import Path
from typing import Union, TYPE_CHECKING

if TYPE_CHECKING:
    from alpm.alpm_types import SchemaVersion

class SourceInfoSchema:
    """SRCINFO schema.

    The schema of a SRCINFO refers to the minimum required sections and keywords, as
    well as the complete set of available keywords in a specific version.
    """

    def __init__(self, version: Union[str, "SchemaVersion"]) -> None:
        """Create a SourceInfoSchema from SchemaVersion or its string representation.

        Args:
            version (Union[str, SchemaVersion]): either a SchemaVersion or a string
                representation of SchemaVersion.

        Raises:
            ALPMError: if the string representation of the version is invalid.
            SourceInfoError: if there is no corresponding schema for the provided major
                version.

        """

    @staticmethod
    def derive_from_str(srcinfo: str) -> "SourceInfoSchema":
        """Derive the schema from a string containing SRCINFO data.

        Args:
            srcinfo (str): The srcinfo string to derive the schema from.

        Returns:
            SourceInfoSchema: The derived schema.

        Raises:
            SourceInfoError: if the srcinfo string cannot be parsed.

        """

    @staticmethod
    def derive_from_file(path: Union[str, Path]) -> "SourceInfoSchema":
        """Derive the schema from a file containing SRCINFO data.

        Args:
            path (Union[str, Path]): The path to the file containing SRCINFO data.

        Returns:
            SourceInfoSchema: The derived schema.

        Raises:
            SourceInfoError: if the file cannot be read or parsed.

        """

    @property
    def version(self) -> "SchemaVersion":
        """The schema version."""

    def __str__(self) -> str: ...
    def __repr__(self) -> str: ...
    def __eq__(self, other: object) -> bool: ...

__all__ = [
    """SourceInfoSchema""",
]
