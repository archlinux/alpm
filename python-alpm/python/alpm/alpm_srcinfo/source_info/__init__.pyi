"""Data representations and integrations for reading of SRCINFO data."""

from . import v1

__all__ = ["v1"]
