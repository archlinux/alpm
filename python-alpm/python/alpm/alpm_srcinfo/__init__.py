# TODO: <https://gitlab.archlinux.org/archlinux/alpm/alpm/-/issues/204>
from alpm import alpm_srcinfo as _mod

__all__ = _mod.__all__
for name in getattr(_mod, "__all__"):
    globals()[name] = getattr(_mod, name)
