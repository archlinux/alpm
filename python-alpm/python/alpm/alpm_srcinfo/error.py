# TODO: <https://gitlab.archlinux.org/archlinux/alpm/alpm/-/issues/204>
from alpm.alpm_srcinfo import error as _mod

__all__ = _mod.__all__
for name in getattr(_mod, "__all__"):
    globals()[name] = getattr(_mod, name)
