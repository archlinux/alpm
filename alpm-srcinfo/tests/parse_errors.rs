use std::{fs::read_to_string, path::PathBuf};

use alpm_srcinfo::SourceInfo;
use insta::assert_snapshot;
use rstest::rstest;
use testresult::TestResult;

/// .SRCINFO files are processed in two steps:
///
/// - Raw parsing step with winnow into intermediate representation (IR)
/// - Bring IR into proper struct representation and apply lints
///
/// This test tests parse errors during the first step.
#[rstest]
pub fn ensure_parse_errors(#[files("tests/parse_errors/*")] case: PathBuf) -> TestResult {
    // Read the input file and parse it.
    let input = read_to_string(&case)?;
    let result = SourceInfo::from_string(input.as_str());

    // Make sure there're no parse errors
    let Err(error) = result else {
        return Err("The parser succeeded even though it should've failed parsing.".into());
    };

    let name = case.file_stem().unwrap().to_str().unwrap();

    // Run the tests with the input being displayed as the description.
    // This makes reviewing this whole stuff a lot easier.
    // Also remove the usual module prefix by explicitly setting the snapshot path.
    // This isn't necessary, as we're already manually sorting snapshots by test scenario.
    let input_clone = input.clone();
    insta::with_settings!({
        description => input_clone,
        snapshot_path => "parse_error_snapshots",
        prepend_module_to_snapshot => false,
    }, {
        assert_snapshot!(name, format!("{error}"));
    });

    Ok(())
}
