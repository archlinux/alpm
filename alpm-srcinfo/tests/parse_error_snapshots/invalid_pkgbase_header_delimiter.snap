---
source: alpm-srcinfo/tests/parse_errors.rs
description: "# Delimiter should have spaces.\npkgbase=test\n    pkgver = 1.0.0\n\tpkgrel = 1\n"
expression: "format!(\"{error}\")"
---
File parsing error:
parse error at line 2, column 8
  |
2 | pkgbase=test
  |        ^
invalid pkgbase section header delimiter
expected ' = '
